### PR TITLE
Add an example with a custom parser (documentation)

### PR DESCRIPTION
**Pull request by xadupre**

### CODE REVIEW

File: docs/examples/plot_custom_parser.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""
+.. _l-custom-parser:
+
+When a custom model has is nor a classifier neither a regressor
```
Comment (prabhat00155) [inline docs/examples/plot_custom_parser.py]: When a custom model is neither a classifier nor a regressor.

File: docs/examples/plot_custom_parser.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+===============================================================
+
+*scikit-learn*'s API specifies a regressor produces one
+outputs, the prediction and a classifier produces two
+outputs, predicted labels and probabilities. In some cases,
+a custom needs to produce different outputs. The goal here is
```
Comment (prabhat00155) [inline docs/examples/plot_custom_parser.py]: You mean 'a custom model' right?

File: docs/examples/plot_custom_parser.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+to add a third results which tells if the probability is
```
Comment (prabhat00155) [inline docs/examples/plot_custom_parser.py]: a third **result**

File: docs/examples/plot_custom_parser.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+above a given threshold which somehow tells the classifier
+valides the answer.
```
Comment (prabhat00155) [inline docs/examples/plot_custom_parser.py]: to validate the result

File: docs/examples/plot_custom_parser.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+.. contents::
+    :local:
+
+Iris and scoring
+++++++++++++++++
+
+To do that, we implement a new estimator which is identical
```
Comment (prabhat00155) [inline docs/examples/plot_custom_parser.py]: indentical to? custom model?

File: docs/examples/plot_custom_parser.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+import onnx
+import sklearn
+from sklearn.base import ClassifierMixin, BaseEstimator, clone
+from sklearn.datasets import load_iris
+from sklearn.linear_model import LogisticRegression
+from sklearn.model_selection import train_test_split
+from skl2onnx import update_registered_converter
+import os
+from onnx.tools.net_drawer import GetPydotGraph, GetOpNodeProducer
+import onnxruntime as rt
```
Comment (prabhat00155) [inline docs/examples/plot_custom_parser.py]: You have imported onnxruntime twice(see line 30).

File: docs/examples/plot_custom_parser_alternative.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+# It fails because the library expected the model
+# to behave like a classifier which produces two
+# outputs. We need to add a custom parser to
+# tell the library this model produces three outputs.
+#
+# Custom parser
+# +++++++++++++
+
+
+def validator_classifier_parser(scope, model, inputs, custom_parsers=None):
```
Comment (prabhat00155) [inline docs/examples/plot_custom_parser_alternative.py]: Where do you use custom_parsers?

File: tests/test_algebra_onnx_operators_sub_estimator.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+                      op_version=opv)
+    r3 = OnnxIdentity(valid, output_names=[outputs[2].full_name],
+                      op_version=opv)
+
+    r1.add_to(scope, container)
+    r2.add_to(scope, container)
+    r3.add_to(scope, container)
+
+
+def validator_classifier_parser(scope, model, inputs, custom_parsers=None):
```
Comment (prabhat00155) [inline tests/test_algebra_onnx_operators_sub_estimator.py]: Unused argument custom_parsers

File: tests/test_algebra_onnx_operators_sub_estimator.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+class TestOnnxOperatorSubEstimator(unittest.TestCase):
+
+    @unittest.skipIf(
+        StrictVersion(ort_version) < StrictVersion("1.0"),
+        reason="Cast not available.")
+    def test_sub_estimator(self):
+        data = load_iris()
+        X, y = data.data, data.target
+        X_train, X_test, y_train, y_test = train_test_split(X, y)
```
Comment (prabhat00155) [inline tests/test_algebra_onnx_operators_sub_estimator.py]: Unused y_test

File: docs/examples/plot_custom_parser.py [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+
+    def validate(self, X):
+        pred = self.predict_proba(X)
+        mx = pred.max(axis=1)
+        return (mx >= self.threshold) * 1
+
+
+data = load_iris()
+X, y = data.data, data.target
+X_train, X_test, y_train, y_test = train_test_split(X, y)
```
Comment (prabhat00155) [inline docs/examples/plot_custom_parser.py]: unused y_test

File: docs/examples/plot_custom_parser_alternative.py [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+
+    def validate(self, X):
+        pred = self.predict_proba(X)
+        mx = pred.max(axis=1)
+        return (mx >= self.threshold) * 1
+
+
+data = load_iris()
+X, y = data.data, data.target
+X_train, X_test, y_train, y_test = train_test_split(X, y)
```
Comment (prabhat00155) [inline docs/examples/plot_custom_parser_alternative.py]: unused y_test

File: docs/examples/plot_custom_parser.py [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+# ++++++++++++++++
+#
+# We reuse some pieces of code from :ref:`l-custom-model`.
+# The shape calculator defines the shape of every output
+# of the converted model.
+
+
+def validator_classifier_shape_calculator(operator):
+
+    input = operator.inputs[0]  # inputs in ONNX graph
```
Comment (prabhat00155) [inline docs/examples/plot_custom_parser.py]: input is a build-in function in Python, I would suggest you to use a different name.

File: docs/examples/plot_custom_parser.py (hunk 10 above)
Comment (xadupre) [reply docs/examples/plot_custom_parser.py]: Unused but any other name would confuse the reader.

Comment (prabhat00155) [reply docs/examples/plot_custom_parser.py]: I mean you coulf use `_`.

File: docs/examples/plot_custom_parser_alternative.py (hunk 11 above)
Comment (prabhat00155) [reply docs/examples/plot_custom_parser_alternative.py]: You may use `_`.

File: docs/examples/plot_custom_parser.py (hunk 12 above)
Comment (xadupre) [reply docs/examples/plot_custom_parser.py]: fixed

File: docs/examples/plot_custom_parser.py (hunk 10 above)
Comment (xadupre) [reply docs/examples/plot_custom_parser.py]: I know but even not used, i still think y_test is better than _.

Comment (prabhat00155) [reply docs/examples/plot_custom_parser.py]: Okay, can you fix the remaining comments?

File: docs/examples/plot_custom_parser.py (hunk 1 above)
Comment (xadupre) [reply docs/examples/plot_custom_parser.py]: fixed

File: docs/examples/plot_custom_parser.py (hunk 2 above)
Comment (xadupre) [reply docs/examples/plot_custom_parser.py]: fixed

File: docs/examples/plot_custom_parser.py (hunk 3 above)
Comment (xadupre) [reply docs/examples/plot_custom_parser.py]: fixed

File: docs/examples/plot_custom_parser.py (hunk 4 above)
Comment (xadupre) [reply docs/examples/plot_custom_parser.py]: fixed

File: tests/test_algebra_onnx_operators_sub_estimator.py (hunk 8 above)
Comment (xadupre) [reply tests/test_algebra_onnx_operators_sub_estimator.py]: All other parsers follow that signature. It is used for sub models.

File: docs/examples/plot_custom_parser.py (hunk 10 above)
Comment (xadupre) [reply docs/examples/plot_custom_parser.py]: I think I did.

Comment (prabhat00155) [reply docs/examples/plot_custom_parser.py]: Look at the comment above, you have imported onnxruntime twice. Also, there are a couple of comments about unused custom_parsers below.

Comment (xadupre) [reply docs/examples/plot_custom_parser.py]: Fixed

**Pull request merged**